### PR TITLE
Add SHA256 signatures of the install script

### PR DESCRIPTION
**Pull request by josegomezr**
#### Proposed Changes ####

Add SHA256 signatures of the install script.

#### Types of Changes ####

- Security fix

#### Verification ####

```bash
sha256sum -c install.sh.sha256sum 
```

#### Testing ####

- No functional change of k3s is needed.

#### Linked Issues ####

* https://github.com/k3s-io/k3s/issues/8447

#### User-Facing Change ####
No functional change

```release-note
- Add SHA256 signatures of the install script.
```

#### Further Comments ####

- Allow install script verification. Downloading artifacts directly from 3rd parties in internet without verification poses a security issue. It's currently impossible to verify the integrity of the installer unless fetching it directly from GitHub. Please provide these signatures alongside the installer and make it available through `get.k3s.io` too.

### CODE REVIEW

**Comment by brandond** [comment]
I'm on the fence about this, as anyone running a `curl | bash` install script has already given up on security, but I guess this doesn't make things any worse.

Please add a test to the validate script to confirm that the checksums match the install script; to avoid accidentally updating the script without touching the checksums.

**Comment by josegomezr** [comment]
@brandond Please have a look.  I've added invocations to `sha{256,384}` in the `./scripts/validate` script, I understood that the script is reached by all CI. Please correct me if I'm wrong.

And may I ask, how can the signatures be part of the release artifacts? what should I edit to have them there too?

---

As a side note:
> As anyone running a `curl | bash` install script has already given up on security.

I partially agree, I see the comfort for a PoC on an isolated machine to do `curl | sh` and get your cluster up and running, but for a production setting a secure option should be advertised too so end users are aware.

It might be silly (and I'm not an expert in security at all) but I like the simple way composer.org offers to use their install guide:

image

Yes, you're downloading things from the wild internet, but you get to verify the integrity of the script with what the docs are telling you (a lot of oversimplification here, I grant that).

Another side-side note, a quick scan over the repo shows: that _technique_ (_curl pipe sh_) is present in:

- Dokerfile.dapper:

https://github.com/k3s-io/k3s/blob/af50e1b09687367979537b835505307a341be818/Dockerfile.dapper#L45C1-L47

- Dokerfile.local:
https://github.com/k3s-io/k3s/blob/af50e1b09687367979537b835505307a341be818/Dockerfile.local#L26-L28

in a Vagrantfile:

- e2e/tailscale/Vagrantfile:
https://github.com/k3s-io/k3s/blob/af50e1b09687367979537b835505307a341be818/tests/e2e/tailscale/Vagrantfile#L31

They don't seem to be a problem now, but are definitely potential attack surfaces, they should be closed in as time permits.

**Comment by brandond** [comment]
> They don't seem to be a problem now, but are definitely potential attack surfaces, they should be closed in as time permits.

That's actually the documented install method for both projects:
* https://golangci-lint.run/usage/install/#other-ci
* https://tailscale.com/kb/1031/install-linux/

**Comment by josegomezr** [comment]
> That's actually the documented install method for both projects:

They do are. That doesn't necessarily means it's a "safe" way to install, it's the same `curl | sh` _technique_ without any sort of check.

I think we're both on the same page: the problem itself is not running `sh`, but rather not being able to verify that the downloaded script is supposed to be the correct script (and if possible, comes from the actual origin, but that needs GPG or something like that).

And what I'd like to achieve is to have that very minimum thing, a way to validate the integrity of the script that's running as the root user of a system and performing

---

Tailscale offers packaged RPM's & DEB's with GPG signatures, so at least is possible to verify the package (which is done automagically by dnf/zypper/apt/etc).

And `golangci-lint` has a recommendation in their own install guide:

image

> **IMPORTANT:** It's highly recommended installing a specific version of golangci-lint available on the [releases page](https://github.com/golangci/golangci-lint/releases).

But then below has a snippet of code `curl | sh`-ing without verification :sweat_smile:

Nevertheless, when following the releases page link, there's  a checksums file to verify the integrity of the artifact downloaded.

File: scripts/validate [hunk 1]
```diff
@@ -8,6 +8,10 @@ fi
 
 cd $(dirname $0)/..
 
+echo Running: install script signature checks
+sha384sum -c install.sh.sha384sum
```
[the remark this answers is not in the document]
Comment (josegomezr) [reply scripts/validate]: Addressed in b68389d0

Kept sha256 sums.

**Review by dereknola — APPROVED** [review]

**Review by brandond — APPROVED** [review]

**Comment by josegomezr** [comment]
Updated the title and description to match the current state (only sha256 signatures).

**Comment by brandond** [comment]
> how can the signatures be part of the release artifacts

Just to close the loop on this - they will not be release artifacts. The install script is served directly off the master branch; it is not part of the built release artifacts. If someone wants to use the checksum, they'll need to pull it down from github.

**Comment by foursixnine** [comment]
> how can the signatures be part of the release artifacts
> 
> Just to close the loop on this - they will not be release artifacts. The install script is served directly off the master branch; it is not part of the built release artifacts. If someone wants to use the checksum, they'll need to pull it down from github.
> 
It is still a step forward, and we can pull the checksum directly from openqa.opensuse.org or openqa.suse.de and I guess security concious users or those working in regulated industries can also refer to it.

**Pull request merged**